### ISSUE TITLE
TIFF with 'new-style' JPEG treated as YCbCr displayed "pink"

### ISSUE THREAD

**Issue by gitdode** [haraldk/TwelveMonkeys#406]
We have some TIFFs with 'new-style' JPEG inside where i.e. white turns to pink.

The cid0 = 0, cid1 = 1 and cid2 = 2, so the colorspace of the JPEG is assumed to be YCbCr, while I think it actually is sRGB - at least the image is displayed correctly with JPEGColorSpace.RGB.

I had a look [here](https://docs.oracle.com/javase/8/docs/api/javax/imageio/metadata/doc-files/jpeg_metadata.html) and from what I understand, there is no assumption made on componend ids 0, 1 and 2 and if subsampled, YCbCr is assumed and if not, RGB is assumed.

In 3.2.2, JPEGImageReader did that, but in the current 3.4-SNAPSHOT it doesn't. Is there any reason for that?

Just to try, I put that check back in 3.4-SNAPSHOT and the image is displayed correctly.

Anything I can contribute with?

**Comment by haraldk** [comment]
Hi Torsten,

The JPEG "color space" detection in the `JPEGImageReader` was rewritten, to closely match that of libJPEG, to reduce the inconsistencies between the Java and the native code. I've tested it on a large set of test data, so I'm pretty confident it's good. 

For TIFF though, the "color space" is explicitly defined by the `PhotometricInterpretation` tag (according to the spec, but of course, some software still sets it wrong; see #404). So we shouldn't have to rely on the JPEG detection algorithm. We can just use the TIFF value, and convert accordingly if needed (maybe with some special tweaks if the TIFF and the JPEG disagrees). 

Can you send me one or two of the TIFFs that displays the problem?

-- 
Harald K

**Comment by gitdode** [comment]
Hi Harald,

Interesting, thanks for the explanation.

Here is an example that shows the problem: 
[s7o_1.tar.gz](https://github.com/haraldk/TwelveMonkeys/files/1720858/s7o_1.tar.gz)

I have a few more but those I would have to email you.

Cheers,
Torsten

**Comment by haraldk** [comment]
Thanks!

Great! This image just broke my attempt at fixing #404... :-P

TL;DR: The old code had two issues, that cancelled each other out, so that the image looked okay.

Long version, there's multiple issues in play here: 

* First, it's the long standing bug #108, which make the metadata report RGB color space (which just happens to be correct in this case, but is generally wrong based on the JPEG stream)
* The TIFF `PhotometricInterpretation` is 1/RGB, and because this is the same as the JPEG metadata says, the `TIFFImageReader` chooses the happy case of leaving color space conversion to the `JPEGImageReader`.
* The new and improved color space detection in the reader reports YCbCr, and happily (and incorrectly) applies its YCbCr to RGB conversion.... 
* Bonus: The TIFF structure has `YCbCrSubsamping` 2,2, however, the JPEG stream is *not* chroma subsampled, which is why the old algorithm got it "right" (the TIFF even has this subsampling setting for the grayscale image in IFD1, which is clearly bogus).


Best regards,

-- 
Harald K

**Comment by haraldk** [comment]
Okay,

I just pushed a fix for this issue to the master branch. 

I've implemented a more reliable way of detecting the JPEG color space (workaround for #108) in TIFF now, which solves the problem for your supplied file.

Please build the latest master, and verify that it fixes the issue for your other files as well. Feel free to reopen the issue if you still have color space issues for "new style" JPEG/7 TIFFs. 


Best regards,

-- 
Harald K

**Comment by gitdode** [comment]
Hi Harald,

I've just tested it and I can confirm that all 3 "pink TIFFs" I currently have now display correctly.

That's great - thanks a lot!

Kind Regards,
Torsten
